### DISCUSSION TITLE
Using a password with special characters in connection string

### DISCUSSION THREAD

**Discussion by tochi-ukpai** [prisma/prisma#15679] [Q&A] [answered]
Hello,
I trying to connect to a db with an @ symbol as part of the password in the connection string.

<img width="627" alt="Screenshot 2022-10-06 at 00 07 32" src="https://user-images.githubusercontent.com/43303739/194179849-4cc9ce72-7f2f-43b1-ac0d-9a3bb4546665.png">

The password is truncated and the rest of the password is read as the db host.

<img width="587" alt="Screenshot 2022-10-06 at 00 09 23" src="https://user-images.githubusercontent.com/43303739/194180049-237b6fdc-beeb-4a69-817f-cfd963969a21.png">

How do I resolve this issue?

**Comment by janpio** [comment, marked as answer]
Use `%40` instead: https://www.prisma.io/docs/reference/database-reference/connection-urls#special-characters

**Reply by tochi-ukpai** [reply]
Hello @janpio 
I've tried that and I got a different error about the db port

<img width="549" alt="Screenshot 2022-10-06 at 01 06 27" src="https://user-images.githubusercontent.com/43303739/194185426-b6abffb1-fbbe-4c89-a097-c7f1df37b712.png">

**Reply by tochi-ukpai** [reply]
I've figured out the issue with the port. I had to also percentage encode "#" in the password. It works fine now.

**Reply by janpio** [reply]
True, that also needs to be encoded indeed 👍

**Reply by DanielHoffmann** [reply]
For anyone who run into this problem, you can uri-encode the password in shellscript using the `jq` command:

```
encodedPassword=$(printf %s "$password" | jq -s -R -r @uri)
dbUrl="postgres://$username:$encodedPassword@$dbAddress/$schemaName"
```

Most environments don't come with `jq` built-in though, so you might need to install it manually in your environment/docker-container

**Reply by Irfan-wani** [reply]
> Use `%40` instead: https://www.prisma.io/docs/reference/database-reference/connection-urls#special-characters

Even after encoding the password i am still getting the same error (I tried URL encoding the whole DB url too, but no luck):
`Error: P1000: Authentication failed against database server, the provided database credentials for `postgres` are not valid.`

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!

**Comment by aleks58andar** [comment]
Hi, here is a function I created which takes care of this issue in Python, if anyone comes across this thread as I did. It specifically deals with the two @'s. It imports the needed methods from urllib.parse, which is a standard Python library. 

    def encode_password_in_db_url(url: str) -> str:
        if isinstance(url, bytes):
            url = url.decode("utf-8")

        parsed = urlparse(url)

        if "@" not in parsed.netloc:
            return url

        try:
            userinfo, _, host = parsed.netloc.rpartition("@")
            if ":" not in userinfo:
                return url

            username, password = userinfo.split(":", 1)
            safe_password = quote_plus(password)

            fixed_netloc = f"{username}:{safe_password}@{host}"
            return urlunparse(parsed._replace(netloc=fixed_netloc))
        except Exception as e:
            raise ValueError(f"Failed to encode password in DB URL: {e}")`

**Comment by jbkkd** [comment]
Docs have been moved here:
https://www.prisma.io/docs/orm/reference/connection-urls#special-characters
